tiledb/sm/query/reader.cc

```diff
@@ -976,15 +976,25 @@ Status Reader::copy_var_cells(
     const auto& var_offsets = var_offsets_per_cs[cs_idx];
 
     // Get tile information, if the range is nonempty.
-    uint64_t* tile_offsets = nullptr;
+    std::vector<uint64_t> tile_offsets;
     Tile* tile_var = nullptr;
     uint64_t tile_cell_num = 0;
     if (cs.tile_ != nullptr) {
-      const auto tile_pair = cs.tile_->tile_pair(name);
-      Tile* const tile = &tile_pair->first;
+      std::pair<tiledb::sm::Tile, tiledb::sm::Tile>* const tile_pair =
+          cs.tile_->tile_pair(name);
+      const auto& tile = &tile_pair->first;
       tile_var = &tile_pair->second;
-      tile_offsets = (uint64_t*)tile->internal_data();
       tile_cell_num = tile->cell_num();
+
+      // Build the tile offsets.
+      tile_offsets.reserve(tile_cell_num);
+      uint64_t tile_offset;
+      uint64_t read_offset = 0;
+      for (size_t cell_idx = 0; cell_idx < tile_cell_num; ++cell_idx) {
```

Comment:
This will probably hurt performance :/. Note that here you are also double-buffering on the offset tile (essentially a full copy into the tile offsets vector).

Reply:
I will revisit this soon to see if there's a better way to do this, same for below. Thanks for pointing this out.

Reply:
The auxiliary offset buffering has been refactored out (although not trivially so).